lib/membrane_rtc_engine/events/track_variant_switched.ex

```diff
@@ -0,0 +1,26 @@
+defmodule Membrane.RTC.Engine.Event.TrackVariantSwitched do
+  @moduledoc """
+  Event sent whenever Engine starts sending a new track variant.
+
+  See also `Membrane.RTC.Engine.Event.SwitchTrackVariant`.
+  """
+
+  alias Membrane.RTC.Engine.Track
+
+  @derive Membrane.EventProtocol
+
+  @typedoc """
+  Type describing TrackVariantSwitched event.
+
+  * `new_variant` - variant that engine will be sending from now
+  * `buffer` - first buffer of a new variant. It's always
+  keyframe (or its beginning).
+  """
+  @type t :: %__MODULE__{
+          new_variant: Track.variant(),
+          buffer: Membrane.Buffer.t()
```

Comment:
I don't understand why the buffer is put inside this event, this seems really off.
1. In your implementation you duplicate that buffer, but it isn't obvious this is the way to go (docs do not mention it)
2. The buffer, as pointed out by the docs, might be just the beginning of the keyframe - which makes me question the usefulness of this even more

Within the context of this PR, it doesn't make sense at all as it isn't used. Could you elaborate on the purpose of this?

Reply:
At the moment buffers are duplicated just to keep backward compatibility. Sending first buffer in an event should make it easier to reconfigure some element states. In other case each element had to have a flag like `awaiting_buffer_from_new_layer` and handling this in `handle_process` i.e.

```elixir
def handle_process(buffer, state) do
 if awaiting_buffer_from new_layer do
   reconfigure(buffer, state)
 else 
   # do casual work
 end
end
```

Reply:
Regarding a keyframe, it's not guaranteed we will receive a keyframe in one RTP packet so a keyframe migth be spread across multiple Membrane buffers

Reply:
> At the moment buffers are duplicated just to keep backward compatibility. Sending first buffer in an event should make it easier to reconfigure some element states. In other case each element had to have a flag like `awaiting_buffer_from_new_layer` and handling this in `handle_process` (...)

Well, then I'm strongly against putting the buffer in the event. First of all, that's against the concepts of the framework, "hiding" the buffer breaks the demands mechanism (or requires additional logic to fix it when receiving such an event). The argument about element reconfiguration doesn't justify that. What kind of elements would require the buffer's payload for reconfiguration? The event should contain the info for reconfiguration, but should not be used to transport the payload.
Besides, it's quite a common pattern to have a special case for the first buffer and the event could be used to trigger it again.

> Regarding a keyframe, it's not guaranteed we will receive it in one RTP packet so it migth be spread across multiple Membrane buffers

I know and that's why I considered this addition of buffer even less useful - trying to figure out the purpose of that I thought the event could be used to store the keyframe to reuse it later on, but that's not possible having only a part of it.